Enhancement: add layout options for email conversion

RazielleS opened the pull request:
## Proposed change

This PR adds the possibility to choose the PDF layout for fetched emails. I tried to not change the current behavior, so it defaults to having Text, then HTML in the PDF.
There now is a dropdown in the email Rules setting, so the behavior can be changed per rule.

![image](https://github.com/user-attachments/assets/1d93fbc7-a36a-44cd-a706-1d120cf88ab1)

It is not possible yet to change the layout for directly uploaded .eml files. I see that it might be another improvement to the PR, but i'm not sure if this is even needed. I might add it if I can find some time.

This implements the changes for #2312.

To test this change:
- Set every possible option in the new dropdown and fetch a new mail
- Also test other documents, since I had to slightly change the consumer

Remaining issues:
- I can't seem to localize the labels, as running `ng extract-i18n` gets me tons of errors. I might need some help from a maintainer here.
- Running `pytest` as well as `npx playwright test` fails quite a bit of tests. But the same seems to happen on the `dev` branch...so maybe something is still missing in my setup (using the vscode devcontainers)
- I have not written any tests, as I have no experience with this in Python.

Please note that I am not usually developing with Python, so the code might be a bit crude in some cases.

Closes #2312.

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:
- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

RazielleS committed:
Feat: Add pdf layout switch for email

shamoon commented:
Im going to mark this as a draft because as noted, there is no testing and the language files aren't updated. But this looks to be how I was thinking to implement this generally speaking, so thanks.

Does your note above imply you do not intend to add the tests?

RazielleS commented:
I'd love to add tests, but I am not even sure if my setup completely works. Furthermore, I do not really have experience with tests in Python.
So if you have any time for it, I would appreciate it if you could add them, if needed.
Otherwise, I will take a look at it, but this might take a lot of time.

shamoon commented on src-ui/src/app/components/common/edit-dialog/mail-rule-edit-dialog/mail-rule-edit-dialog.component.ts:
```diff
@@ -58,6 +59,25 @@ const CONSUMPTION_SCOPE_OPTIONS = [
+    name: $localize`HTML, then text`,
+  },
+  {
+    id: MailRulePdfLayout.Html_only,
```
```suggestion
    id: MailRulePdfLayout.HtmlOnly,
```

RazielleS replied:
Done

shamoon commented on src-ui/src/app/components/common/edit-dialog/mail-rule-edit-dialog/mail-rule-edit-dialog.component.ts:
```diff
@@ -58,6 +59,25 @@ const CONSUMPTION_SCOPE_OPTIONS = [
+    name: $localize`HTML only`,
+  },
+  {
+    id: MailRulePdfLayout.Text_only,
```
```suggestion
    id: MailRulePdfLayout.TextOnly,
```

RazielleS replied:
Done

shamoon committed:
Update frontend translation strings

shamoon commented on src-ui/src/app/components/common/edit-dialog/mail-rule-edit-dialog/mail-rule-edit-dialog.component.ts:
```diff
@@ -58,6 +59,25 @@ const CONSUMPTION_SCOPE_OPTIONS = [
 
+const PDF_LAYOUT_OPTIONS = [
+  {
+    id: MailRulePdfLayout.Text_Html,
```
```suggestion
    id: MailRulePdfLayout.TextHtml,
```

RazielleS replied:
Done

shamoon commented on src-ui/src/app/components/common/edit-dialog/mail-rule-edit-dialog/mail-rule-edit-dialog.component.ts:
```diff
@@ -58,6 +59,25 @@ const CONSUMPTION_SCOPE_OPTIONS = [
+    name: $localize`Text, then HTML`,
+  },
+  {
+    id: MailRulePdfLayout.Html_Text,
```
```suggestion
    id: MailRulePdfLayout.HtmlText,
```

RazielleS replied:
Done

shamoon commented on src-ui/src/app/data/mail-rule.ts:
```diff
@@ -11,6 +11,13 @@ export enum MailRuleConsumptionScope {
 }
 
+export enum MailRulePdfLayout {
+  Text_Html = 1,
+  Html_Text = 2,
+  Html_only = 3,
+  Text_only = 4,
```
```suggestion
  TextHtml = 1,
  HtmlText = 2,
  HtmlOnly = 3,
  TextOnly = 4,
```

RazielleS replied:
Done

shamoon commented on src/paperless_mail/parsers.py:
```diff
@@ -121,7 +122,13 @@ def extract_metadata(self, document_path: Path, mime_type: str):
+        document_path: Path,
+        mime_type: str,
+        file_name=None,
+        mailrule: int | None = None,
```
```suggestion
        mailrule_id: int | None = None,
```

RazielleS replied:
Done

shamoon commented on src/paperless_mail/parsers.py:
```diff
@@ -180,7 +187,11 @@ def build_formatted_text(mail_message: MailMessage) -> str:
 
         self.log.debug("Creating a PDF from the email")
-        self.archive_path = self.generate_pdf(mail)
+        if mailrule:
+            rule = MailRule.objects.get(pk=mailrule)
```
```suggestion
        if mailrule_id:
            rule = MailRule.objects.get(pk=mailrule_id)
```

RazielleS replied:
Done

shamoon committed:
Fix broken test

shamoon reviewed (changes requested):
Just some nits. I can look at the tests soon, it shouldn't be much

shamoon commented on src/paperless_mail/parsers.py:
```diff
@@ -246,7 +261,17 @@ def generate_pdf(self, mail_message: MailMessage) -> Path:
 
-                route.merge([mail_pdf_file, pdf_of_html_content])
+                match pdf_layout:
+                    case MailRule.PdfLayout.TEXT_HTML:
+                        route.merge([mail_pdf_file, pdf_of_html_content])
+                    case MailRule.PdfLayout.HTML_TEXT:
+                        route.merge([pdf_of_html_content, mail_pdf_file])
+                    case MailRule.PdfLayout.HTML_ONLY:
+                        route.merge([pdf_of_html_content])
+                    case MailRule.PdfLayout.TEXT_ONLY:
+                        route.merge([mail_pdf_file])
+                    case _:
+                        route.merge([mail_pdf_file, pdf_of_html_content])
```
```suggestion
                    case MailRule.PdfLayout.HTML_TEXT:
                        route.merge([pdf_of_html_content, mail_pdf_file])
                    case MailRule.PdfLayout.HTML_ONLY:
                        route.merge([pdf_of_html_content])
                    case MailRule.PdfLayout.TEXT_ONLY:
                        route.merge([mail_pdf_file])
                    case MailRule.PdfLayout.TEXT_HTML | _:
                        route.merge([mail_pdf_file, pdf_of_html_content])
```

RazielleS replied:
Done

SyberSycho commented:
This seems like an odd way to implement this request. Since as you mentioned it only would effect .eml when imported using a mail rule. If you ever need to reparse the .eml I think it would default to the normal way unless you reimported through the mail rule. It is definitely an improvement however compared to no option.

I had a lightbulb moment when thinking about what was said here https://github.com/paperless-ngx/paperless-ngx/discussions/2312#discussioncomment-10980060. I also had some free time at home so I tried my hand at coding a different solution.

Now I don't really know how to set things up for testing with what I coded however and don't have the time to figure it out. I also have barely touched python. So with all that being said I think the below code would work for all .eml files regardless of how it was imported into paperless.

Still need code for the actually option which would be better located in the application settings since it should work on all .eml files.

Things that I surrounded with *** should be changed to an appropriate variable

Code would replace the def build_formatted_text at https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/paperless_mail/parsers.py#L138
```
def build_formatted_text(mail_message: MailMessage) -> str:
    """
    Constructs a formatted string, based on the given email.
    Basically tries to get most of the email content & included front matter into a nice string
    """
    fmt_text = f"Subject: {mail_message.subject}\n\n"
    fmt_text += f"From: {mail_message.from_values.full}\n\n"
    to_list = [address.full for address in mail_message.to_values]
    fmt_text += f"To: {', '.join(to_list)}\n\n"
    if mail_message.cc_values:
        fmt_text += (
            f"CC: {', '.join(address.full for address in mail.cc_values)}\n\n"
        )
    if mail_message.bcc_values:
        fmt_text += (
            f"BCC: {', '.join(address.full for address in mail.bcc_values)}\n\n"
        )
    if mail_message.attachments:
        att = []
        for a in mail.attachments:
            attachment_size = naturalsize(a.size, binary=True, format="%.2f")
            att.append(
                f"{a.filename} ({attachment_size})",
            )
        fmt_text += f"Attachments: {', '.join(att)}\n\n"

    # Does not keep text if EML_PDF_OPTION is set to HTML_ONLY, unless no HTML content is present
    if not mail.html and mail.text: # If there is no HTML content and there is text content
        fmt_text += f"\n\n{strip_text(mail.text)}"
    elif mail.html and ***EML_PDF_OPTION = HTML_ONLY***: # If there is HTML content and no text content wanted
        fmt_text += "HTML content: " + strip_text(self.tika_parse(mail.html))
    else # If both html and text content is wanted
        if mail.text: # If there is no text content it will not be included. Hopefully this would help avoid any blank PDFs sections
            fmt_text += f"\n\n{strip_text(mail.text)}"
        fmt_text += "HTML content: " + strip_text(self.tika_parse(mail.html))

    return fmt_text
```

Code would replace the def generate_pdf at https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/paperless_mail/parsers.py#L220
```
def generate_pdf(self, mail_message: MailMessage) -> Path:
archive_path = Path(self.tempdir) / "merged.pdf"

# If no HTML content, create the PDF from the message
# Otherwise, create 2 PDFs and merge them with Gotenberg
# Order of text and HTML content in the final PDF is determined by the user's settings
if not mail_message.html: # No HTML content
    archive_path.write_bytes(self.generate_pdf_from_mail(mail_message))
elif: mail_message.html and ***EML_PDF_OPTION = HTML_ONLY***: # HTML content only
    archive_path.write_bytes(self.generate_pdf_from_html(mail_message.html, mail_message.attachments))
else:
    mail_pdf_file = self.generate_pdf_from_mail(mail_message)
    pdf_of_html_content = self.generate_pdf_from_html(mail_message.html, mail_message.attachments)

    self.log.debug("Merging email text and HTML content into single PDF")

    with (
        GotenbergClient(
            host=settings.TIKA_GOTENBERG_ENDPOINT,
            timeout=settings.CELERY_TASK_TIME_LIMIT,
        ) as client,
        client.merge.merge() as route,
    ):
        # Configure requested PDF/A formatting, if any
        pdf_a_format = self._settings_to_gotenberg_pdfa()
        if pdf_a_format is not None:
            route.pdf_format(pdf_a_format)

        if: ***EML_PDF_OPTION = HTML_FIRST***: # HTML content first
            route.merge([pdf_of_html_content, mail_pdf_file])
        else: # TEXT content first
            route.merge([mail_pdf_file, pdf_of_html_content])

        try:
            response = route.run()
            archive_path.write_bytes(response.content)
        except Exception as err:
            raise ParseError(
                f"Error while merging email HTML into PDF: {err}",
            ) from err

return archive_path
```

shamoon commented:
> This seems like an odd way to implement this request.

I completely disagree. even just looking at how compact the change is suggests it's the right way. As the OP also noted, the changes here could be made to support generic .eml parsing but that is not the primary use case here.

SyberSycho commented:
If I knew how to use github better and show just the changes my solution would actually show to be more compact, while not limiting the function to only new .eml thought the mail rules.

Like I said in my previous post however. I'm not really much of a coder and don't really have a good way of testing nor the time for what I linked above. It really doesn't matter however. I mainly just linked what I did for a better coder than myself to look over it and maybe they would see it as a good implementation, even if the code might need touch ups to ensure it is working code.

shamoon committed:
Setting for .eml files consumed outside of mail fetch

shamoon commented:
https://github.com/paperless-ngx/paperless-ngx/pull/8907/commits/303847847fbdb8581ec1963378cd079e7c876149 should allow controlling .eml parsing outside of mail fetch

RazielleS committed:
Implement suggestions in PR

RazielleS committed:
Feat: Add default PDF layout option as fallback to ENV

RazielleS commented:
@shamoon I changed everything according to the suggestions you made.
After you added the ENV variable for global usage, I think it makes sense that a default is available in the dropdown.
So I went ahead and added this new value to the new dropdown. When selecting it, it will just use whatever is the default, and fall back to Text, HTML if nothing is set at all.
(Also this time, the language file generation worked :tada:)

shamoon commented on src/paperless_mail/models.py:
```diff
@@ -230,6 +237,12 @@ class CorrespondentSource(models.IntegerChoices):
+    pdf_layout = models.PositiveIntegerField(
+        _("pdf layout"),
+        choices=PdfLayout.choices,
+        default=PdfLayout.TEXT_HTML,
```
Shouldn't this use the new `DEFAULT`?

shamoon committed:
Add coverage

RazielleS commented:
You are absolutely correct. I have missed this one and will change it when I'm back home. Thank you!

Am 26. Januar 2025 16:20:17 MEZ schrieb shamoon ***@***.***>:
>@shamoon commented on this pull request.
>
>
>
>> @@ -230,6 +237,12 @@ class CorrespondentSource(models.IntegerChoices):
>         default=ConsumptionScope.ATTACHMENTS_ONLY,
>     )
> 
>+    pdf_layout = models.PositiveIntegerField(
>+        _("pdf layout"),
>+        choices=PdfLayout.choices,
>+        default=PdfLayout.TEXT_HTML,
>
>Shouldn't this use the new `DEFAULT`?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/paperless-ngx/paperless-ngx/pull/8907#pullrequestreview-2574310005
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

shamoon commented on src/documents/consumer.py:
```diff
@@ -474,7 +475,18 @@ def progress_callback(current_progress, max_progress):  # pragma: no cover
             )
             self.log.debug(f"Parsing {self.filename}...")
-            document_parser.parse(self.working_copy, mime_type, self.filename)
+            if (
+                isinstance(document_parser, MailDocumentParser)
+                and self.input_doc.mailrule_id
+            ):
+                document_parser.parse(
+                    self.working_copy,
+                    mime_type,
+                    self.filename,
+                    self.input_doc.mailrule_id,
+                )
+            else:
+                document_parser.parse(self.working_copy, mime_type, self.filename)
```
```suggestion
            document_parser.parse(
                document_path=self.working_copy,
                mime_type=mime_type,
                file_name=self.filename,
                mailrule_id=self.input_doc.mailrule_id,
            )
```

I'm trying to think if we need to pass the rule id here (or can get it later) but either way, just pass this along without the if/else, `self.input_doc.mailrule_id` defaults to None, so I *think* this will work.

shamoon replied:
Actually, come to think of it. Changing the declaration of parser.parse is sort of problematic, it would need to be changed everywhere. You have circumvented this with `isinstance(document_parser, MailDocumentParser)`

Need to think about this one but I think we need a better solution than this (and not my suggestion)

shamoon replied:
@stumpylog any thoughts on this question? (no rush)

RazielleS replied:
Yes this is exactly why I used `isinstance`. Without it, I broke the other passers in my first try.

shamoon replied:
Yea, but this feels like a hack, or a change that will affect a lot of code. I'll think on it

shamoon replied:
mostly asking stumpylog, how about this: e3119507bbe96262339868cbbc13eb2b795ae7bf dunno if that's a little hack-y but otherwise I dont see a way besides altering the parser declaration (or leaving as it currently is).

Again, no rush at all.

stumpylog replied:
Something I have been thinking about is passing parser specific settings to the parser.  So I would be fine altering the declaration of `parse` with the other ones not using it (yet)

shamoon replied:
Great thanks stumpy. This PR is probably ready, for whenever we start merging our next round of enhancements

shamoon commented on src/documents/consumer.py:
```diff
@@ -47,6 +47,7 @@
 from documents.utils import copy_basic_file_stats
 from documents.utils import copy_file_with_basic_stats
 from documents.utils import run_subprocess
+from paperless_mail.parsers import MailDocumentParser
```
```suggestion
```

RazielleS committed:
Bugfix: Change default layout in model

shamoon committed:
Cover the consumer change

shamoon committed:
Revert / fix lang file changes

shamoon committed:
Merge branch 'dev' into pr/8907

stumpylog reviewed (approved):
I'm good with this as it.  I might do some iteration on how the rule is passed to the parser, but this is good for now

stumpylog committed:
Merge branch 'dev' into 2312-add-pdf-layout-choice

stumpylog committed:
Merge branch 'dev' into 2312-add-pdf-layout-choice
